gpkit/nomial_data.py

```diff
@@ -0,0 +1,128 @@
+
+from functools import reduce as functools_reduce
+from operator import add
+
+from .small_classes import HashVector, Quantity
+
+from .small_scripts import mag
+
+
+class NomialData(object):
+
+    def __init__(self, exps=None, cs=None, signomials=None, simplify=True):
+        if signomials:
+            exps = functools_reduce(add, (s.exps for s in signomials))
+            cs = np.hstack((mag(s.cs) for s in signomials))
+        if exps is None or cs is None:
+            raise ValueError("creation of a NomialData requires exps and cs.")
+        if simplify:
+            exps, cs = sort_and_simplify(exps, cs)
+        self.exps, self.cs = exps, cs
+        self.varlocs, self.varstrs = locate_vars(self.exps)
+        self.values = {vk: vk.descr["value"] for vk in self.varlocs
+                       if "value" in vk.descr}
+        self._hash = hash(tuple(zip(self.exps, tuple(self.cs))))
+
+    def __hash__(self):
+        return self._hash
+
+    def __repr__(self):
+        return "gpkit.%s(%s)" % (self.__class__.__name__, str(self._hash))
+
+
+def sort_and_simplify(exps, cs, return_map=False):
+    """Reduces the number of monomials, and casts them to a sorted form.
+
+    Arguments
+    ---------
+
+    exps : list of Hashvectors
+        The exponents of each monomial
+    cs : array of floats or Quantities
+        The coefficients of each monomial
+    return_map : bool (optional)
+        Whether to return the map of which monomials combined to form a
+        simpler monomial, and their fractions of that monomial's final c.
+
+    Returns
+    -------
+
+    exps : list of Hashvectors
+        Exponents of simplified monomials.
+    cs : array of floats or Quantities
+        Coefficients of simplified monomials.
+    mmap : list of tuples
+        List for each original monomial of (destination index, fraction)
+    """
+    matches = defaultdict(float)
+    if return_map:
+        expmap = defaultdict(dict)
+    for i, exp in enumerate(exps):
+        exp = HashVector({var: x for (var, x) in exp.items() if x != 0})
+        matches[exp] += cs[i]
+        if return_map:
+            expmap[exp][i] = cs[i]
+
+    if len(matches) > 1:
+        zeroed_terms = (exp for exp, c in matches.items() if mag(c) == 0)
+        for exp in zeroed_terms:
+            del matches[exp]
+
+    exps_ = tuple(matches.keys())
+    cs_ = list(matches.values())
+    if isinstance(cs_[0], Quantity):
+        units = cs_[0]/cs_[0].magnitude
+        cs_ = [c.to(units).magnitude for c in cs_] * units
+    else:
+        cs_ = np.array(cs_, dtype='float')
+
+    if not return_map:
+        return exps_, cs_
+    else:
+        mmap = [None]*len(cs)
+        for i, item in enumerate(matches.items()):
+            exp, c = item
+            for j in expmap[exp]:
+                mmap[j] = (i, expmap[exp][j]/c)
+        return exps_, cs_, mmap
+
+
+def locate_vars(exps):
+    "From exponents form a dictionary of which monomials each variable is in."
```

Comment:
I haven't carefully examined the logic in this method  -- should I look through it carefully?